### DISCUSSION TITLE
Running generators as schematics is deprecated warning when composing an nx generator?

### DISCUSSION THREAD

**Discussion by kiwidude68** [nrwl/nx#17980] [General]
I've been wanting to compose the @nrwl/angular:library generator (using Nx 15.9.2) to do a bunch of things (e.g. default some values based on a library subtype, removing the auto-generated standalone component, adding in some folders of our own etc). 

Everything is working great EXCEPT I get this warning every time my generator runs:
`Running generators as schematics is deprecated and will be removed in v17. Prefer ``callRule(convertNxGenerator(generator)(options), tree, context)`` where "generator" is the name of the generator you wish to use.
`

This happens when in my generator the following code (snippet) is invoked, which looks "similar" the Compose docs page:
https://nx.dev/plugins/recipes/composing-generators

```
import { libraryGenerator } from '@nrwl/angular/generators';

export default async function (tree: Tree, options: LibraryGeneratorSchema) {

  // Generate a library using the Nrwl plugin with our own set of options
  await libraryGenerator(tree, {
    ...options,
   <extra parameters>
  }
  ... do other stuff ...
}
```

My own generator in schema.json has "cli": "nx", and if I omit invoking the nrwl library generator then the warning disappears. So it is something about invoking the nrwl library generator from within another generator?

Any help/advice would be welcomed - it is really annoying noise for our developers to see and I don't know how to avoid it? Is this specific to using the particular versions of Angular 15/Nx we are currently on and will disappear once we move to 16.x or ?

**Comment by kiwidude68** [comment]
If it helps at all, it seems like the issue is where it loads the generators.ts file, and somehow invokes schematic warnings for the convert-ts-lint-to-eslint and move generators (neither of which I directly use, only libraryGenerator), as shown by this jest test output below?
```
PS D:\Git\my-project> node "node_modules/jest/bin/jest.js" "d:/Git/my-project/libs/plugins/src/generators/p1-library/p1-library.generator.spec.ts" -c "d:/Git/my-project/libs/plugins/jest.config.ts" -t "p1-library generator should remove default standalone component"
  console.warn
    Running generators as schematics is deprecated and will be removed in v17. Prefer `callRule(convertNxGenerator(generator)(options), tree, context)` where "generator" is the name of the generator you wish to use.

      at warnForSchematicUsage (../../packages/angular/src/generators/utils/warn-for-schematic-usage.ts:2:11)
      at Object.<anonymous> (../../packages/angular/src/generators/convert-tslint-to-eslint/convert-tslint-to-eslint.ts:141:57)
      at Object.<anonymous> (../../packages/angular/generators.ts:6:1)
      at Object.<anonymous> (src/generators/p1-library/p1-library.generator.ts:6:1)
      at Object.<anonymous> (src/generators/p1-library/p1-library.generator.spec.ts:6:1)

  console.warn
    Running generators as schematics is deprecated and will be removed in v17. Prefer `callRule(convertNxGenerator(generator)(options), tree, context)` where "generator" is the name of the generator you wish to use.

      at warnForSchematicUsage (../../packages/angular/src/generators/utils/warn-for-schematic-usage.ts:2:11)
      at Object.<anonymous> (../../packages/angular/src/generators/move/move.ts:28:58)
      at Object.<anonymous> (../../packages/angular/generators.ts:12:1)
      at Object.<anonymous> (src/generators/p1-library/p1-library.generator.ts:6:1)
      at Object.<anonymous> (src/generators/p1-library/p1-library.generator.spec.ts:6:1)
```

Looking at the `generators.json` inside the @nrwl/angular branch for 15.9.x I can see that all the generators get exposed through some kind of compat.ts file that wraps convertToNxGenerator with warnings. However I'm just trying to import the exported libraryGenerator function, just like the code examples do so don't understand why this all happens.

Is anyone successfully using composed libraryGenerator function in their own generators, and avoiding these console.warn messages somehow? What is the secret - is 15.9.x the problem here? Is there a simple example libraryGenerator extended repo out there which doesn't have this problem I could look at? The documentation on generators is very lacking, as are isolated simple examples...

**Reply by SairamPotta** [reply]
After upgrading from nx 13 to 15 we are also getting the same warning message. Any updates on this warning messages on how to solve

**Reply by kiwidude68** [reply]
As you can see no-one bothered replying to the thread so I'm still in the dark on this one too unfortunately.

**Comment by leosvelperez** [comment]
This was fixed by https://github.com/nrwl/nx/pull/16396 and the change was released in Nx v16. Please note that if you're composing a generator (and not the compat schematic) there's no real issue other than the incorrect warning. If you can update to Nx v16, the warning will go away.
